File: src_algorighm/triangulate/triangulate.py
from math import isclose, sqrt


class Point:
    def __init__(self, x=0, y=0, z=0):
        self.x = x
        self.y = y
        self.z = z

    def int_2d(self):
        return [int(self.x), int(self.y)]


class Edge:
    def __init__(self):
        self.p1 = 0
        self.p2 = 0


class Triangle:
    def __init__(self):
        self.p1 = 0
        self.p2 = 0
        self.p3 = 0


def circumCircle(xp, yp, x1, y1, x2, y2, x3, y3, circle):
    """
    Return ture if pos is inside the circumCircle made up of the triangle
    the circumCircle center pos and the radius ard also returned

    """

    if isclose(y1, y2) and isclose(y2, y3):
        return False

    if isclose(y2, y1):
        m2 = - (x3 - x2) / (y3 - y2)
        mx2 = (x2 + x3) / 2.0
        my2 = (y2 + y3) / 2.0
        xc = (x2 + x1) / 2.0
        yc = m2 * (xc - mx2) + my2
    elif isclose(y3, y2):
        m1 = - (x2 - x1) / (y2 - y1)
        mx1 = (x1 + x2) / 2.0
        my1 = (y1 + y2) / 2.0
        xc = (x3 + x2) / 2.0
        yc = m1 * (xc - mx1) + my1
    else:
        m1 = - (x2 - x1) / (y2 - y1)
        m2 = - (x3 - x2) / (y3 - y2)
        mx1 = (x1 + x2) / 2.0
        mx2 = (x2 + x3) / 2.0
        my1 = (y1 + y2) / 2.0
        my2 = (y2 + y3) / 2.0
        xc = (m1 * mx1 - m2 * mx2 + my2 - my1) / (m1 - m2)
        yc = m1 * (xc - mx1) + my1

    dx = x2 - xc
    dy = y2 - yc
    rsqr = dx * dx + dy * dy
    r = sqrt(rsqr)

    dx = xp - xc
    dy = yp - yc
    drsqr = dx * dx + dy * dy

    circle[0:3] = [xc, yc, r]
    return drsqr <= rsqr
    pass


def triangulate(points):
    """
    Triangulation subroutine
    Takes as input NV vertices in array points
    Returned is a list of ntri triangular
    """
    nv = len(points) - 3
    assert nv > 0
    nedge = 0
    emax = 200
    trimax = 4 * nv
    complete = [False for x in range(trimax)]
    edges = [Edge() for x in range(emax)]
    v = [Triangle() for x in range(3 * nv)]
    # sort pxyz

    # Find the maximum and minimum vertex bounds.
    # This is to allow calculation of the bounding triangle
    xmin = points[0].x
    ymin = points[0].y
    xmax = xmin
    ymax = ymin
    for i in range(1, nv):
        if points[i].x < xmin:
            xmin = points[i].x
        if points[i].x > xmax:
            xmax = points[i].x
        if points[i].y < ymin:
            ymin = points[i].y
        if points[i].y > ymax:
            ymax = points[i].y

    dx = xmax - xmin
    dy = ymax - ymin
    dmax = max(dx, dy)
    xmid = (xmax + xmin) / 2.0
    ymid = (ymax + ymin) / 2.0

    """
      Set up the supertriangle
      This is a triangle which encompasses all the sample points.
      The supertriangle coordinates are added to the end of the
      vertex list. The supertriangle is the first triangle in
      the triangle list.
    """
    points[nv + 0].x = xmid - 20 * dmax
    points[nv + 0].y = ymid - dmax
    points[nv + 0].z = 0.0
    points[nv + 1].x = xmid
    points[nv + 1].y = ymid + 20 * dmax
    points[nv + 1].z = 0.0
    points[nv + 2].x = xmid + 20 * dmax
    points[nv + 2].y = ymid - dmax
    points[nv + 2].z = 0.0
    v[0].p1 = nv
    v[0].p2 = nv + 1
    v[0].p3 = nv + 2
    complete[0] = False
    ntri = 1

    # Include each point one at a time into the existing mesh
    for i in range(nv):
        xp = points[i].x
        yp = points[i].y
        nedge = 0
        """
         Set up the edge buffer.
         If the point (xp,yp) lies inside the circumcircle then the
         three edges of that triangle are added to the edge buffer
         and that triangle is removed.
         """
        circle = [0, 0, 0]
        j = 0
        while j < ntri:
            if complete[j]:
                j += 1
                continue
            x1 = points[v[j].p1].x
            y1 = points[v[j].p1].y
            x2 = points[v[j].p2].x
            y2 = points[v[j].p2].y
            x3 = points[v[j].p3].x
            y3 = points[v[j].p3].y
            inside = circumCircle(xp, yp, x1, y1, x2, y2, x3, y3, circle)
            xc = circle[0]
            yc = circle[1]
            r = circle[2]
            if xc + r < xp:
                complete[j] = True
            if inside:
                if nedge + 3 >= emax:
                    emax += 100
                    for ie in range(100):
                        edges.append(Edge())
                edges[nedge + 0].p1 = v[j].p1
                edges[nedge + 0].p2 = v[j].p2
                edges[nedge + 1].p1 = v[j].p2
                edges[nedge + 1].p2 = v[j].p3
                edges[nedge + 2].p1 = v[j].p3
                edges[nedge + 2].p2 = v[j].p1
                nedge += 3
                v[j] = v[ntri - 1]
                complete[j] = complete[ntri - 1]
                ntri -= 1
                j -= 1
            j += 1
        """
         Tag multiple edges
         Note: if all triangles are specified anticlockwise then all
               interior edges are opposite pointing in direction.
        """
        for j in range(nedge - 1):
            for k in range(j + 1, nedge):
                if (edges[j].p1 == edges[k].p2) and (edges[j].p2 == edges[k].p1):
                    edges[j].p1 = -1
                    edges[j].p2 = -1
                    edges[k].p1 = -1
                    edges[k].p2 = -1
                # Shouldn't need the following, see note above 
                if (edges[j].p1 == edges[k].p1) and (edges[j].p2 == edges[k].p2):
                    edges[j].p1 = -1
                    edges[j].p2 = -1
                    edges[k].p1 = -1
                    edges[k].p2 = -1
        """
         Form new triangles for the current point
         Skipping over any tagged edges.
         All edges are arranged in clockwise order.
        """
        for j in range(nedge):
            if edges[j].p1 < 0 or edges[j].p2 < 0:
                continue
            if ntri >= trimax:
                raise Exception("ntri >= trimax")
            v[ntri].p1 = edges[j].p1
            v[ntri].p2 = edges[j].p2
            v[ntri].p3 = i
            complete[ntri] = False
            ntri += 1
    """
    Remove triangles with supertriangle vertices
    These are triangles which have a vertex number greater than nv
    """
    for i in range(ntri):
        if v[i].p1 >= nv or v[i].p2 >= nv or v[i].p3 >= nv:
            v[i] = v[ntri - 1]
            ntri -= 1
            i -= 1
    return v[0:ntri]
    pass

File: algorighm/triangulate/triangulate.c
typedef struct {
   int p1,p2,p3;
} ITRIANGLE;
typedef struct {
   int p1,p2;
} IEDGE;
typedef struct {
   double x,y,z;
} XYZ;

/*
   Triangulation subroutine
   Takes as input NV vertices in array pxyz
   Returned is a list of ntri triangular faces in the array v
   These triangles are arranged in a consistent clockwise order.
   The triangle array 'v' should be malloced to 3 * nv
   The vertex array pxyz must be big enough to hold 3 more points
   The vertex array must be sorted in increasing x values say

   qsort(p,nv,sizeof(XYZ),XYZCompare);
      :
   int XYZCompare(void *v1,void *v2)
   {
      XYZ *p1,*p2;
      p1 = v1;
      p2 = v2;
      if (p1->x < p2->x)
         return(-1);
      else if (p1->x > p2->x)
         return(1);
      else
         return(0);
   }
*/
int Triangulate(int nv,XYZ *pxyz,ITRIANGLE *v,int *ntri)
{
   int *complete = NULL;
   IEDGE *edges = NULL;
   int nedge = 0;
   int trimax,emax = 200;
   int status = 0;

   int inside;
   int i,j,k;
   double xp,yp,x1,y1,x2,y2,x3,y3,xc,yc,r;
   double xmin,xmax,ymin,ymax,xmid,ymid;
   double dx,dy,dmax;

   /* Allocate memory for the completeness list, flag for each triangle */
   trimax = 4 * nv;
   if ((complete = malloc(trimax*sizeof(int))) == NULL) {
      status = 1;
      goto skip;
   }

   /* Allocate memory for the edge list */
   if ((edges = malloc(emax*(long)sizeof(EDGE))) == NULL) {
      status = 2;
      goto skip;
   }

   /*
      Find the maximum and minimum vertex bounds.
      This is to allow calculation of the bounding triangle
   */
   xmin = pxyz[0].x;
   ymin = pxyz[0].y;
   xmax = xmin;
   ymax = ymin;
   for (i=1;i<nv;i++) {
      if (pxyz[i].x < xmin) xmin = pxyz[i].x;
      if (pxyz[i].x > xmax) xmax = pxyz[i].x;
      if (pxyz[i].y < ymin) ymin = pxyz[i].y;
      if (pxyz[i].y > ymax) ymax = pxyz[i].y;
   }
   dx = xmax - xmin;
   dy = ymax - ymin;
   dmax = (dx > dy) ? dx : dy;
   xmid = (xmax + xmin) / 2.0;
   ymid = (ymax + ymin) / 2.0;

   /*

   */
   pxyz[nv+0].x = xmid - 20 * dmax;
   pxyz[nv+0].y = ymid - dmax;
   pxyz[nv+0].z = 0.0;
   pxyz[nv+1].x = xmid;
   pxyz[nv+1].y = ymid + 20 * dmax;
   pxyz[nv+1].z = 0.0;
   pxyz[nv+2].x = xmid + 20 * dmax;
   pxyz[nv+2].y = ymid - dmax;
   pxyz[nv+2].z = 0.0;
   v[0].p1 = nv;
   v[0].p2 = nv+1;
   v[0].p3 = nv+2;
   complete[0] = FALSE;
   *ntri = 1;

   /*
      Include each point one at a time into the existing mesh
   */
   for (i=0;i<nv;i++) {

      xp = pxyz[i].x;
      yp = pxyz[i].y;
      nedge = 0;

      /*
         Set up the edge buffer.
         If the point (xp,yp) lies inside the circumcircle then the
         three edges of that triangle are added to the edge buffer
         and that triangle is removed.
      */
      for (j=0;j<(*ntri);j++) {
         if (complete[j])
            continue;
         x1 = pxyz[v[j].p1].x;
         y1 = pxyz[v[j].p1].y;
         x2 = pxyz[v[j].p2].x;
         y2 = pxyz[v[j].p2].y;
         x3 = pxyz[v[j].p3].x;
         y3 = pxyz[v[j].p3].y;
         inside = CircumCircle(xp,yp,x1,y1,x2,y2,x3,y3,&xc,&yc,&r);
         if (xc < xp && ((xp-xc)*(xp-xc)) > r)
				complete[j] = TRUE;
         if (inside) {
            /* Check that we haven't exceeded the edge list size */
            if (nedge+3 >= emax) {
               emax += 100;
               if ((edges = realloc(edges,emax*(long)sizeof(EDGE))) == NULL) {
                  status = 3;
                  goto skip;
               }
            }
            edges[nedge+0].p1 = v[j].p1;
            edges[nedge+0].p2 = v[j].p2;
            edges[nedge+1].p1 = v[j].p2;
            edges[nedge+1].p2 = v[j].p3;
            edges[nedge+2].p1 = v[j].p3;
            edges[nedge+2].p2 = v[j].p1;
            nedge += 3;
            v[j] = v[(*ntri)-1];
            complete[j] = complete[(*ntri)-1];
            (*ntri)--;
            j--;
         }
      }

      /*
         Tag multiple edges
         Note: if all triangles are specified anticlockwise then all
               interior edges are opposite pointing in direction.
      */
      for (j=0;j<nedge-1;j++) {
         for (k=j+1;k<nedge;k++) {
            if ((edges[j].p1 == edges[k].p2) && (edges[j].p2 == edges[k].p1)) {
               edges[j].p1 = -1;
               edges[j].p2 = -1;
               edges[k].p1 = -1;
               edges[k].p2 = -1;
            }
            /* Shouldn't need the following, see note above */
            if ((edges[j].p1 == edges[k].p1) && (edges[j].p2 == edges[k].p2)) {
               edges[j].p1 = -1;
               edges[j].p2 = -1;
               edges[k].p1 = -1;
               edges[k].p2 = -1;
            }
         }
      }

      /*
         Form new triangles for the current point
         Skipping over any tagged edges.
         All edges are arranged in clockwise order.
      */
      for (j=0;j<nedge;j++) {
         if (edges[j].p1 < 0 || edges[j].p2 < 0)
            continue;
         if ((*ntri) >= trimax) {
            status = 4;
            goto skip;
         }
         v[*ntri].p1 = edges[j].p1;
         v[*ntri].p2 = edges[j].p2;
         v[*ntri].p3 = i;
         complete[*ntri] = FALSE;
         (*ntri)++;
      }
   }

   /*
      Remove triangles with supertriangle vertices
      These are triangles which have a vertex number greater than nv
   */
   for (i=0;i<(*ntri);i++) {
      if (v[i].p1 >= nv || v[i].p2 >= nv || v[i].p3 >= nv) {
         v[i] = v[(*ntri)-1];
         (*ntri)--;
         i--;
      }
   }

skip:
   free(edges);
   free(complete);
   return(status);
}

/*
   Return TRUE if a point (xp,yp) is inside the circumcircle made up
   of the points (x1,y1), (x2,y2), (x3,y3)
   The circumcircle centre is returned in (xc,yc) and the radius r
   NOTE: A point on the edge is inside the circumcircle
*/
int CircumCircle(double xp,double yp,
   double x1,double y1,double x2,double y2,double x3,double y3,
   double *xc,double *yc,double *rsqr)
{
   double m1,m2,mx1,mx2,my1,my2;
   double dx,dy,drsqr;
   double fabsy1y2 = fabs(y1-y2);
   double fabsy2y3 = fabs(y2-y3);

   /* Check for coincident points */
   if (fabsy1y2 < EPSILON && fabsy2y3 < EPSILON)
       return(FALSE);

   if (fabsy1y2 < EPSILON) {
      m2 = - (x3-x2) / (y3-y2);
      mx2 = (x2 + x3) / 2.0;
      my2 = (y2 + y3) / 2.0;
      *xc = (x2 + x1) / 2.0;
      *yc = m2 * (*xc - mx2) + my2;
   } else if (fabsy2y3 < EPSILON) {
      m1 = - (x2-x1) / (y2-y1);
      mx1 = (x1 + x2) / 2.0;
      my1 = (y1 + y2) / 2.0;
      *xc = (x3 + x2) / 2.0;
      *yc = m1 * (*xc - mx1) + my1;
   } else {
      m1 = - (x2-x1) / (y2-y1);
      m2 = - (x3-x2) / (y3-y2);
      mx1 = (x1 + x2) / 2.0;
      mx2 = (x2 + x3) / 2.0;
      my1 = (y1 + y2) / 2.0;
      my2 = (y2 + y3) / 2.0;
      *xc = (m1 * mx1 - m2 * mx2 + my2 - my1) / (m1 - m2);
      if (fabsy1y2 > fabsy2y3) {
         *yc = m1 * (*xc - mx1) + my1;
      } else {
         *yc = m2 * (*xc - mx2) + my2;
      }
   }

   dx = x2 - *xc;
   dy = y2 - *yc;
   *rsqr = dx*dx + dy*dy;

   dx = xp - *xc;
   dy = yp - *yc;
   drsqr = dx*dx + dy*dy;

   // Original
   //return((drsqr <= *rsqr) ? TRUE : FALSE);
   // Proposed by Chuck Morris
   return((drsqr - *rsqr) <= EPSILON ? TRUE : FALSE);
}

/* Crappy code to test triangulation */

int XYZCompare(void *,void *);

int main(int argc,char **argv)
{
   int i;
   int ntri = 0;
   double x,y,z;
   ITRIANGLE *v;
   XYZ *p = NULL;
   int nv = 0;
   FILE *fptr;

   if (argc < 2) {
      fprintf(stderr,"usage: %s datafile\n",argv[0]);
      exit(-1);
   }
   if ((fptr = fopen(argv[1],"r")) == NULL) {
      fprintf(stderr,"Failed to open file \"%s\"\n",argv[1]);
      exit(-1);
   }
   while (fscanf(fptr,"%lf %lf %lf",&x,&y,&z) == 3) {
      p = realloc(p,(nv+1)*sizeof(XYZ));
      p[nv].x = x;
      p[nv].y = y;
      p[nv].z = z;
      nv++;
   }
   fprintf(stderr,"Read %d points\n",nv);
   if (nv < 3)
      exit(-1);
   fclose(fptr);

   p = realloc(p,(nv+3)*sizeof(XYZ));
   v = malloc(3*nv*sizeof(TRIANGLE));
   qsort(p,nv,sizeof(XYZ),XYZCompare);
   Triangulate(nv,p,v,&ntri);
   fprintf(stderr,"Formed %d triangles\n",ntri);

   /* Write triangles in geom format */
   for (i=0;i<ntri;i++) {
      printf("f3 %g %g %g %g %g %g %g %g %g 1 1 1\n",
         p[v[i].p1].x,p[v[i].p1].y,p[v[i].p1].z,
         p[v[i].p2].x,p[v[i].p2].y,p[v[i].p2].z,
         p[v[i].p3].x,p[v[i].p3].y,p[v[i].p3].z);
   }
   exit(0);
}

int XYZCompare(void *v1,void *v2)
{
   XYZ *p1,*p2;
   p1 = v1;
   p2 = v2;
   if (p1->x < p2->x)
      return(-1);
   else if (p1->x > p2->x)
      return(1);
   else
      return(0);
}

